[WPE] Implement non-composited page rendering

Scony opened the pull request:
#### f47827c7c3aa15647991c6ec44a94ae3552e35f2
<pre>
[WPE] Implement non-composited page rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=303170">https://bugs.webkit.org/show_bug.cgi?id=303170</a>

Reviewed by Carlos Garcia Campos.

This change implements the non-composited page rendering for WPE.

With this change, it&apos;s possible to utilize a simplified frame rendering
that does not use compositor at all. It&apos;s useful especially on low-end
embedded devices where GPU is not powerful or not present at all.

The non-composited mode implemented in this change can be enabled using
a -AcceleratedCompositing preference.

Additionally, this change adds a layout ref-test that protects the new
implementation as well as existing, GTK&apos;s implementation from
regressions.

Canonical link: <a href="https://commits.webkit.org/304678@main">https://commits.webkit.org/304678@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07c56e294097fe9e5084ef3d0f8887ffa0084596

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136257 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143967 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104184 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6756 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85015 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6412 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4072 "Found 2 new API test failures: TestWebKitAPI.MessagePort.MessageToClosedPort, TestWebKitAPI.WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginAndEndEditingEvents (failure)") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4560 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115701 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8296 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40863 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112524 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6968 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112867 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28647 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6340 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118400 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8344 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36457 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8062 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71903 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8136 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
I guess you need to make it current

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
This second nullptr can be omitted

Scony replied:
it is already (6 passed out of 7). With 5 we get: `too few arguments to function 'sk_sp<SkSurface> SkSurfaces::WrapBackendRenderTarget(GrRecordingContext*, const GrBackendRenderTarget&, GrSurfaceOrigin, SkColorType, sk_sp<SkColorSpace>, const SkSurfaceProps*, RenderTargetReleaseProc, ReleaseContext)'`

carlosgcampos replied:
Ah!, first nullptr is the color space.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
I guess we could do this only as a fallback when creating a fence is not possible, no?

Scony replied:
Actually we need to do this first as this one flushes skia internals (surfaces, caches etc.) and eventually it triggers actual drawing.

See:
```
#5  0x00007cd710d8697f in GrGLFunction<void (unsigned int, unsigned int, unsigned int, int, unsigned int, void const*)>::GrGLFunction(void (*)(unsigned int, unsigned int, unsigned int, int, unsigned int, void const*))::{lambda(void const*, unsigned int, unsigned int, unsigned int, int, unsigned int, void const*)#1}::operator()(void const*, unsigned int, unsigned int, unsigned int, int, unsigned int, void const*) const
    (__closure=0x0, buf=0x5be3f6003b18, args#0=4, args#1=0, args#2=19, args#3=30, args#4=5123, args#5=0x0) at ./Source/ThirdParty/skia/include/gpu/ganesh/gl/GrGLFunctions.h:274
#6  0x00007cd710d869db in GrGLFunction<void (unsigned int, unsigned int, unsigned int, int, unsigned int, void const*)>::GrGLFunction(void (*)(unsigned int, unsigned int, unsigned int, int, unsigned int, void const*))::{lambda(void const*, unsigned int, unsigned int, unsigned int, int, unsigned int, void const*)#1}::_FUN(void const*, unsigned int, unsigned int, unsigned int, int, unsigned int, void const*) () at ./Source/ThirdParty/skia/include/gpu/ganesh/gl/GrGLFunctions.h:273
#7  0x00007cd7107ce567 in GrGLFunction<void (unsigned int, unsigned int, unsigned int, int, unsigned int, void const*)>::operator()(unsigned int, unsigned int, unsigned int, int, unsigned int, void const*) const
    (this=0x5be3f6003b10, args#0=4, args#1=0, args#2=19, args#3=30, args#4=5123, args#5=0x0) at ./Source/ThirdParty/skia/include/gpu/ganesh/gl/GrGLFunctions.h:297
#8  0x00007cd7107cd4bd in GrGLOpsRenderPass::onDrawIndexed (this=0x5be3f600c200, indexCount=30, baseIndex=0, minIndexValue=0, maxIndexValue=19, baseVertex=0) at ./Source/ThirdParty/skia/src/gpu/ganesh/gl/GrGLOpsRenderPass.cpp:224
#9  0x00007cd7106e7ff2 in GrOpsRenderPass::drawIndexed (this=0x5be3f600c200, indexCount=30, baseIndex=0, minIndexValue=0, maxIndexValue=19, baseVertex=0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrOpsRenderPass.cpp:246
#10 0x00007cd710898d9f in skgpu::ganesh::QuadPerEdgeAA::IssueDraw (caps=..., renderPass=0x5be3f600c200, spec=..., runningQuadCount=0, quadsInDraw=5, maxVerts=20, absVertBufferOffset=0) at ./Source/ThirdParty/skia/src/gpu/ganesh/ops/QuadPerEdgeAA.cpp:545
#11 0x00007cd71086becb in (anonymous namespace)::FillRectOpImpl::onExecute (this=0x5be3f684a010, flushState=0x7ffda4328ff0, chainBounds=...) at ./Source/ThirdParty/skia/src/gpu/ganesh/ops/FillRectOp.cpp:348
#12 0x00007cd71086e063 in GrOp::execute (this=0x5be3f684a010, state=0x7ffda4328ff0, chainBounds=...) at ./Source/ThirdParty/skia/src/gpu/ganesh/ops/GrOp.cpp:64
#13 0x00007cd71088cf4b in skgpu::ganesh::OpsTask::onExecute (this=0x5be3f68c3030, flushState=0x7ffda4328ff0) at ./Source/ThirdParty/skia/src/gpu/ganesh/ops/OpsTask.cpp:666
#14 0x00007cd7106b85aa in GrRenderTask::execute (this=0x5be3f68c3030, flushState=0x7ffda4328ff0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrRenderTask.h:53
#15 0x00007cd7106ba28d in GrDrawingManager::executeRenderTasks (this=0x5be3f45305c0, flushState=0x7ffda4328ff0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDrawingManager.cpp:290
#16 0x00007cd7106b9f8f in GrDrawingManager::flush (this=0x5be3f45305c0, proxies=..., access=SkSurfaces::BackendSurfaceAccess::kNoAccess, info=..., newState=0x0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDrawingManager.cpp:209
#17 0x00007cd7106bb021 in GrDrawingManager::flushSurfaces (this=0x5be3f45305c0, proxies=..., access=SkSurfaces::BackendSurfaceAccess::kNoAccess, info=..., newState=0x0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDrawingManager.cpp:540
#18 0x00007cd7106b6749 in GrDirectContextPriv::flushSurfaces (this=0x7ffda432ac08, proxies=..., access=SkSurfaces::BackendSurfaceAccess::kNoAccess, info=..., newState=0x0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDirectContextPriv.cpp:92
#19 0x00007cd71066f62d in GrDirectContextPriv::flushSurface (this=0x7ffda432ac08, proxy=0x5be3f0cde930, access=SkSurfaces::BackendSurfaceAccess::kNoAccess, info=..., newState=0x0) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDirectContextPriv.h:106
#20 0x00007cd710670bf2 in GrDirectContext::flush (this=0x5be3f5fbd5c0, surface=0x5be3f600a800, access=SkSurfaces::BackendSurfaceAccess::kNoAccess, info=...) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDirectContext.cpp:502
#21 0x00007cd710670d64 in GrDirectContext::flushAndSubmit (this=0x5be3f5fbd5c0, surface=0x5be3f600a800, sync=GrSyncCpu::kNo) at ./Source/ThirdParty/skia/src/gpu/ganesh/GrDirectContext.cpp:525
#22 0x00007cd7108e62e0 in skgpu::ganesh::FlushAndSubmit (surface=...) at ./Source/ThirdParty/skia/src/gpu/ganesh/surface/SkSurface_Ganesh.cpp:788
#23 0x00007cd70579e128 in WebKit::AcceleratedSurface::RenderTargetShareableBuffer::sync (this=0x7cd6e1019140, useExplicitSync=false) at ./Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:257
```

carlosgcampos replied:
Right, we always need to flush, but if fences are not available we need to flush with GrSyncCpu::kYes. I'm not sure it's equivalent to the glFlush we are doing below, but for now I think it's ok. I would use flushAndSubmit() GrContext method, though, because you have to pass the GrSyncCpu and it's clear we are passing kNo. I had to find skgpu::ganesh::FlushAndSubmit() in the sources to check if it does cpu sync or not.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.h:
you can simply use RELEASE_ASSERT_NOT_REACHED()

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
This is wpe only for now, isn't it?

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
I think it's better to remove this conditionals. We can just define the non composited frame renderer for all ports but only create it for wpe for now.

Scony replied:
eventually those were kept so that we don't have to provide dummy class implementations for win and playstation

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
why 20_us?

Scony replied:
the idea here was to have something bigger than 0 yet reasonably small - is there a better way to do it? or should I just name the constant?

Scony replied:
removed eventually

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
Since this can fail, I think we should add create() method that can return nullptr and then use a private constructor that receives the surface and context. This way the caller can fallback to non-composited renderer without gl (at least when this is used by gtk). Or maybe we want to move here the non-gl renderer code too and fallback internally.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.h:
This should be WeakPtr or RefPtr

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.h:
```suggestion
    bool m_canRenderNextFrame { true };
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
We need to call AcceleratedSurface::willDestroyCompositingRunLoop to remove the message receiver. Also  AcceleratedSurface::willDestroyGLContext() to release gl resources before the context is destroyed.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
Maybe we could do this with skia api after creating the canvas? I think skia can optimize clear for some drivers.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
This should receive a plain reference

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
And this too, and the WeakPtr is created on construction

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
makeUnique

Scony replied:
the constructor in private in this case

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
Now I see why you need the setter for the frame complete. I think we could do something similar to what GraphicsContextGL does, create always allocates the instance, then calls a private initialize() methods that returns a bool and if it fails it returns nullptr and the instance is destroyed.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/NonCompositedFrameRenderer.cpp:
m_context can't be nullptr.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.cpp:
```diff
@@ -181,6 +190,50 @@ void AcceleratedSurface::RenderTargetShareableBuffer::didRenderFrame(Vector<IntR
+#else
+    constexpr auto notImplemented = false;
+    RELEASE_ASSERT(notImplemented);
+    return nullptr;
```
RELEASE_ASSERT_NOT_REACHED()

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/DrawingAreaCoordinatedGraphics.cpp:
```diff
@@ -61,6 +61,10 @@
 #include "LayerTreeHostTextureMapper.h"
 #endif
 
+#if PLATFORM(GTK) || PLATFORM(WPE)
```
```suggestion
#if PLATFORM(WPE)
```

Scony committed:
[WPE] Implement non-composited page rendering
https://bugs.webkit.org/show_bug.cgi?id=303170

Reviewed by Carlos Garcia Campos.

This change implements the non-composited page rendering for WPE.

With this change, it's possible to utilize a simplified frame rendering
that does not use compositor at all. It's useful especially on low-end
embedded devices where GPU is not powerful or not present at all.

The non-composited mode implemented in this change can be enabled using
a -AcceleratedCompositing preference.

Additionally, this change adds a layout ref-test that protects the new
implementation as well as existing, GTK's implementation from
regressions.

Canonical link: https://commits.webkit.org/304678@main
